File: src/third_iteration/override_methods/RockPaperScissorsSolver.java
package third_iteration.override_methods;

import third_iteration.override_methods.Gesture.*;

public class RockPaperScissorsSolver {

	public static <T extends Gesture> String solve(T leftGesture, T rightGesture) {
		return "Tie.";
	}

	public static String solve(Rock leftGesture, Paper rightGesture) {
		return "Rock loses against paper.";
	}

	public static String solve(Rock leftGesture, Scissors rightGesture) {
		return "Rock wins against scissors.";
	}

	public static String solve(Paper leftGesture, Rock rightGesture) {
		return "Paper wins against rock.";
	}

	public static String solve(Paper leftGesture, Scissors rightGesture) {
		return "Paper loses against scissors.";
	}

	public static String solve(Scissors leftGesture, Rock rightGesture) {
		return "Scissors lose against rock.";
	}

	public static String solve(Scissors leftGesture, Paper rightGesture) {
		return "Scissors win against paper.";
	}
}

File: php/src/third_iteration/override_methods/RockPaperScissorsSolver.php
<?php

namespace third_iteration\override_methods;

class RockPaperScissorsSolver
{
    /**
     * @param Gesture $leftGesture
     * @param Gesture $rightGesture
     * @return string
     */
    public static function solve($leftGesture, $rightGesture)
    {
        // PHP does not support overloading so we have to do the mapping to the right method.
        $methodName = 'solve' . self::getShortClassName($leftGesture) . 'Against' . self::getShortClassName($rightGesture);
        return self::$methodName();
    }

    /**
     * @return string
     */
    private static function solveRockAgainstRock()
    {
        return "Tie.";
    }

    /**
     * @return string
     */
    private static function solveRockAgainstPaper()
    {
        return "Rock loses against paper.";
    }

    /**
     * @return string
     */
    private static function solveRockAgainstScissors()
    {
        return "Rock wins against scissors.";
    }

    /**
     * @return string
     */
    private static function solvePaperAgainstRock()
    {
        return "Paper wins against rock.";
    }

    /**
     * @return string
     */
    private static function solvePaperAgainstPaper()
    {
        return "";
    }

    /**
     * @return string
     */
    private static function solvePaperAgainstScissors()
    {
        return "Paper loses against scissors.";
    }

    /**
     * @return string
     */
    private static function solveScissorsAgainstRock()
    {
        return "Scissors lose against rock.";
    }

    /**
     * @return string
     */
    private static function solveScissorsAgainstPaper()
    {
        return "Scissors win against paper.";
    }

    /**
     * @return string
     */
    private static function solveScissorsAgainstScissors()
    {
        return "";
    }

    /**
     * @param $object
     * @return string
     */
    private
    static function getShortClassName($object)
    {
        $path = explode('\\', get_class($object));
        return array_pop($path);
    }
}
